### PR TITLE
Python3 builtin changes

### PR DESCRIPTION
**Pull request by QuLogic**
Several builtins have changed names or semantics and require a bit of special handling.
 * `x.next()` -> `next(x)`
 * `div` -> `truediv`, `__and__` -> `and_`
 * nested contexts
 * `namedtuple.__init__` -> `namedtuple.__new__`
 * Rename of builtin package
 * Formatting of floats (see SciTools/cartopy#426)
 * `mock` -> `unittest.mock`

### CODE REVIEW

File: lib/iris/analysis/maths.py [hunk 1]
```diff
@@ -332,7 +332,10 @@ def divide(cube, other, dim=None, in_place=False):
     _assert_is_cube(cube)
     other_unit = getattr(other, 'units', '1')
     new_unit = cube.units / other_unit
-    op = operator.idiv if in_place else operator.div
+    try:
+        op = operator.itruediv if in_place else operator.truediv
+    except AttributeError:
+        op = operator.idiv if in_place else operator.div
     return _binary_op_common(op, 'divison', cube, other, new_unit, dim,
```
Comment (QuLogic) [inline lib/iris/analysis/maths.py]: Just noticed there's a typo here; I don't know what that second parameter is used for though.

Comment (rhattersley) [inline lib/iris/analysis/maths.py]: What's the typo? :confused:

Comment (ajdawson) [inline lib/iris/analysis/maths.py]: divison -> division

File: lib/iris/analysis/maths.py [hunk 2]
```diff
@@ -332,7 +332,10 @@ def divide(cube, other, dim=None, in_place=False):
     _assert_is_cube(cube)
     other_unit = getattr(other, 'units', '1')
     new_unit = cube.units / other_unit
-    op = operator.idiv if in_place else operator.div
+    try:
+        op = operator.itruediv if in_place else operator.truediv
+    except AttributeError:
+        op = operator.idiv if in_place else operator.div
```
Comment (rhattersley) [inline lib/iris/analysis/maths.py]: 1. I don't see why an AttributeError would ever get raised - Python 2 and 3 both define `operator.[i]truediv`.
2. Changing from `[i]div` to `[i]truediv` is a change of behaviour for Python 2.

Comment (rhattersley) [inline lib/iris/analysis/maths.py]: NB. (1) applies to the other changes in this commit.

Comment (QuLogic) [inline lib/iris/analysis/maths.py]: Yea, I think you're right; I probably swapped them accidentally, because Python 3 definitely doesn't have `[i]div`.

**Comment by QuLogic** [comment]
WRT the floats, this is the method we ended up using in Cartopy. Using `repr` does seem to work for what I've tried so far as well, but you should also check on 32-bit.

File: lib/iris/analysis/maths.py (hunk 1 above)
Comment (rhattersley) [inline lib/iris/analysis/maths.py]: Thanks @ajdawson :smile:

Comment (rhattersley) [inline lib/iris/analysis/maths.py]: > I don't know what that second parameter is used for though.

It's used to customise various errors messages in `_assert_matching_units` and `_broadcast_cube_coord_data`. From the error that begins "Could not determine dimension..." it looks like the best value for this string would be the name of the function, i.e. "divide".

Comment (rhattersley) [inline lib/iris/analysis/maths.py]: > divison -> division

I suggest I fix that in a separate PR.

Comment (rhattersley) [inline lib/iris/analysis/maths.py]: See #1775.

**Comment by rhattersley** [comment]
This PR represents a big chunk of work - thanks @QuLogic! :smile:

**Pull request merged**